pennylane_lightning/core/src/simulators/lightning_kokkos/gates/BasicGateFunctors.hpp

```diff
@@ -1138,7 +1326,71 @@ void applyToffoli(Kokkos::View<Kokkos::complex<PrecisionT> *> arr_,
         });
 }
 
-template <class PrecisionT, class FuncT> class applyNC4Functor {
+template <class PrecisionT, class FuncT, bool has_controls>
+class applyNC4Functor {};
+
+template <class PrecisionT, class FuncT>
+class applyNC4Functor<PrecisionT, FuncT, true> {
+    using KokkosIntVector = Kokkos::View<std::size_t *>;
+
+    Kokkos::View<Kokkos::complex<PrecisionT> *> arr;
+    const FuncT core_function;
+    KokkosIntVector indices;
+    KokkosIntVector parity;
+    KokkosIntVector rev_wires;
+    KokkosIntVector rev_wire_shifts;
+
+  public:
+    template <class ExecutionSpace>
+    applyNC4Functor([[maybe_unused]] ExecutionSpace exec,
+                    Kokkos::View<Kokkos::complex<PrecisionT> *> arr_,
+                    std::size_t num_qubits,
+                    const std::vector<std::size_t> &controlled_wires,
+                    const std::vector<bool> &controlled_values,
+                    const std::vector<std::size_t> &wires, FuncT core_function_)
+        : arr(arr_), core_function(core_function_) {
+
+        std::tie(parity, rev_wires) =
+            Util::reverseWires(num_qubits, wires, controlled_wires);
+        std::vector<std::size_t> indices_ =
+            generateBitPatterns(wires, num_qubits);
+        ControlBitPatterns(indices_, num_qubits, controlled_wires,
+                           controlled_values);
+        indices = vector2view(indices_);
+        Kokkos::parallel_for(
+            Kokkos::RangePolicy<ExecutionSpace>(
+                0, exp2(num_qubits - controlled_wires.size() - wires.size())),
+            *this);
+    }
+    KOKKOS_FUNCTION void operator()(const std::size_t k) const {
+        const std::size_t offset = Util::parity_2_offset(parity, k);
+        const std::size_t i0000 = indices(0B0000);
+        const std::size_t i0001 = indices(0B0001);
+        const std::size_t i0010 = indices(0B0010);
+        const std::size_t i0011 = indices(0B0011);
+        const std::size_t i0100 = indices(0B0100);
+        const std::size_t i0101 = indices(0B0101);
+        const std::size_t i0110 = indices(0B0110);
+        const std::size_t i0111 = indices(0B0111);
+        const std::size_t i1000 = indices(0B1000);
+        const std::size_t i1001 = indices(0B1001);
+        const std::size_t i1010 = indices(0B1010);
+        const std::size_t i1011 = indices(0B1011);
+        const std::size_t i1100 = indices(0B1100);
+        const std::size_t i1101 = indices(0B1101);
+        const std::size_t i1110 = indices(0B1110);
+        const std::size_t i1111 = indices(0B1111);
```

Comment:
Why do we need to define all these variables? Could not be better to create a `Kokkos::view`? As we discussed offline we are concerned about the overhead of each option(many args or create a vector) What do you suggest @maliasadi @AmintorDusko @multiphaseCFD ?

Reply:
For multi-qubit gates, this approach clearly describes the underlying unitary matrix for testing and debugging reasons and has a minimal overhead comparing to alternatives.

Reply:
@LuisAlfredoNu, we have benchmarked a vector-like alternative and explicitly variables like this in the past, verifying that this is a more performant solution.